From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martin Finkel <me@martinfinkel.com>
Date: Mon, 1 Jun 2026 19:00:00 +0200
Subject: [PATCH] win32: dirs: stub WinRTSHGetFolderPath in UWP builds

Newer llvm-mingw no longer declares the WinRT interfaces (IStorageFolder,
IApplicationDataStatics, IApplicationData, IKnownFoldersStatics, etc.)
used in the WinRTSHGetFolderPath body when WINAPI_FAMILY is set to
WINAPI_FAMILY_APP, so src/win32/dirs.c fails to compile in UWP builds.

The winstring.h / windows.storage.h / roapi.h headers are still present
but the types are guarded with WINAPI_FAMILY_PARTITION_DESKTOP and become
undeclared. Replace the body with a stub returning E_NOTIMPL so
config_GetShellDir yields NULL in UWP, matching the existing
config_GetLibDir UWP behavior of returning NULL.
---
 src/win32/dirs.c | 125 +++---------------------------------------------------
 1 file changed, 4 insertions(+), 121 deletions(-)

diff --git a/src/win32/dirs.c b/src/win32/dirs.c
index ef1718f3f3..98125b03ef 100644
--- a/src/win32/dirs.c
+++ b/src/win32/dirs.c
@@ -48,10 +48,6 @@
 #include <limits.h>
 
 #if VLC_WINSTORE_APP
-#include <winstring.h>
-#include <windows.storage.h>
-#include <roapi.h>
-
 #ifndef CSIDL_LOCAL_APPDATA
 # define CSIDL_LOCAL_APPDATA 0x001C
 #endif
@@ -60,123 +56,10 @@ static HRESULT WinRTSHGetFolderPath(HWND hwnd, int csidl, HANDLE hToken, DWORD d
 {
     VLC_UNUSED(hwnd);
     VLC_UNUSED(hToken);
-
-    HRESULT hr;
-    IStorageFolder *folder;
-
-    if (dwFlags != SHGFP_TYPE_CURRENT)
-        return E_NOTIMPL;
-
-    folder = NULL;
-    csidl &= ~0xFF00; /* CSIDL_FLAG_MASK */
-
-    if (csidl == CSIDL_APPDATA) {
-        IApplicationDataStatics *appDataStatics = NULL;
-        IApplicationData *appData = NULL;
-        static const WCHAR *className = L"Windows.Storage.ApplicationData";
-        const UINT32 clen = wcslen(className);
-
-        HSTRING hClassName = NULL;
-        HSTRING_HEADER header;
-        hr = WindowsCreateStringReference(className, clen, &header, &hClassName);
-        if (FAILED(hr))
-            goto end_appdata;
-
-        hr = RoGetActivationFactory(hClassName, &IID_IApplicationDataStatics, (void**)&appDataStatics);
-
-        if (FAILED(hr))
-            goto end_appdata;
-
-        if (!appDataStatics) {
-            hr = E_FAIL;
-            goto end_appdata;
-        }
-
-        hr = IApplicationDataStatics_get_Current(appDataStatics, &appData);
-
-        if (FAILED(hr))
-            goto end_appdata;
-
-        if (!appData) {
-            hr = E_FAIL;
-            goto end_appdata;
-        }
-
-        hr = IApplicationData_get_LocalFolder(appData, &folder);
-
-end_appdata:
-        WindowsDeleteString(hClassName);
-        if (appDataStatics)
-            IApplicationDataStatics_Release(appDataStatics);
-        if (appData)
-            IApplicationData_Release(appData);
-    }
-    else
-    {
-        IKnownFoldersStatics *knownFoldersStatics = NULL;
-        static const WCHAR *className = L"Windows.Storage.KnownFolders";
-        const UINT32 clen = wcslen(className);
-
-        HSTRING hClassName = NULL;
-        HSTRING_HEADER header;
-        hr = WindowsCreateStringReference(className, clen, &header, &hClassName);
-        if (FAILED(hr))
-            goto end_other;
-
-        hr = RoGetActivationFactory(hClassName, &IID_IKnownFoldersStatics, (void**)&knownFoldersStatics);
-
-        if (FAILED(hr))
-            goto end_other;
-
-        if (!knownFoldersStatics) {
-            hr = E_FAIL;
-            goto end_other;
-        }
-
-        switch (csidl) {
-        case CSIDL_PERSONAL:
-            hr = IKnownFoldersStatics_get_DocumentsLibrary(knownFoldersStatics, &folder);
-            break;
-        case CSIDL_MYMUSIC:
-            hr = IKnownFoldersStatics_get_MusicLibrary(knownFoldersStatics, &folder);
-            break;
-        case CSIDL_MYPICTURES:
-            hr = IKnownFoldersStatics_get_PicturesLibrary(knownFoldersStatics, &folder);
-            break;
-        case CSIDL_MYVIDEO:
-            hr = IKnownFoldersStatics_get_VideosLibrary(knownFoldersStatics, &folder);
-            break;
-        default:
-            hr = E_NOTIMPL;
-        }
-
-end_other:
-        WindowsDeleteString(hClassName);
-        if (knownFoldersStatics)
-            IKnownFoldersStatics_Release(knownFoldersStatics);
-    }
-
-    if( SUCCEEDED(hr) && folder != NULL )
-    {
-        HSTRING path = NULL;
-        IStorageItem *item = NULL;
-        PCWSTR pszPathTemp;
-        hr = IStorageFolder_QueryInterface(folder, &IID_IStorageItem, (void**)&item);
-        if (FAILED(hr))
-            goto end_folder;
-        hr = IStorageItem_get_Path(item, &path);
-        if (FAILED(hr))
-            goto end_folder;
-        pszPathTemp = WindowsGetStringRawBuffer(path, NULL);
-        wcscpy(pszPath, pszPathTemp);
-end_folder:
-        WindowsDeleteString(path);
-        IStorageFolder_Release(folder);
-        if (item)
-            IStorageItem_Release(item);
-    }
-
-    return hr;
+    VLC_UNUSED(csidl);
+    VLC_UNUSED(dwFlags);
+    VLC_UNUSED(pszPath);
+    return E_NOTIMPL;
 }
 #define SHGetFolderPathW WinRTSHGetFolderPath
 #endif
-- 
2.40.0
